proof/courier.go

```diff
@@ -295,53 +359,297 @@ func (h *HashMailCourier) DeliverProof(ctx context.Context, addr address.Taro,
+}
+
+// BackoffExecError is an error returned when the backoff execution fails.
+// This error wraps the underlying error returned by the execution function.
+// It allows the porter to determine whether the state machine should be halted
+// or not.
+type BackoffExecError struct {
+	execErr error
+}
+
+func (e *BackoffExecError) Error() string {
+	return fmt.Sprintf("backoff exec error: %s", e.execErr.Error())
+}
+
+// backoffExec attempts to execute the given `exec` function using a repeating
+// backoff time delayed strategy. The backoff strategy is used to ensure
+// that we don't spam the hashmail service with proof delivery attempts.
+func (h *HashMailCourier) backoffExec(ctx context.Context,
+	exec func() error) error {
+
+	var (
+		backoff          = h.cfg.BackoffCfg.InitialBackoff
+		numTries         = h.cfg.BackoffCfg.NumTries
+		maxBackoff       = h.cfg.BackoffCfg.MaxBackoff
+		err        error = nil
+	)
+
+	for i := 0; i < numTries; i++ {
+		// Don't delay at first try. Wait before reattempting
+		// delivery if a backoff duration is specified.
+		if i > 0 && backoff != 0 {
+			// Notify subscribers that backoff wait is about to
+			// commence.
+			transferEvent := NewReceiverProofBackoffWaitEvent(
+				backoff, int64(i),
+			)
+			h.publishSubscriberEvent(transferEvent)
+
+			err = h.wait(ctx, backoff)
+			if err != nil {
+				return err
+			}
+		}
+
+		err = exec()
+		if err != nil {
+			err = fmt.Errorf("error executing backoff procedure: "+
+				"%w", &BackoffExecError{execErr: err})
+
+			// Function execution wasn't successful, increase the
+			// time we'll wait until the next try.
+			backoff *= 2
+
+			// Cap the backoff at the maximum backoff.
+			if backoff > maxBackoff {
+				backoff = maxBackoff
+			}
+			log.Debugf("Receiver proof delivery failed with "+
+				"error. Backing off for %s: %v", backoff, err)
+
+			if i < numTries-1 {
+				log.Infof("Receiver proof delivery failed, "+
+					"will try again after backoff delay: "+
+					"%v", backoff)
+			}
+
+			// Ready to retry proof delivery if applicable.
+			continue
+		}
+
+		// If the proof delivery was successful, we can exit the loop.
+		break
+	}
+	if err != nil {
+		return fmt.Errorf("receiver proof delivery failed; count "+
+			"retries attempted: %d; %w", numTries, err)
+	}
+
+	return nil
+}
+
+// publishSubscriberEvent publishes an event to all subscribers.
+func (h *HashMailCourier) publishSubscriberEvent(event chanutils.Event) {
+	// Lock the subscriber mutex to ensure that we don't modify the
+	// subscriber map while we're iterating over it.
+	h.subscriberMtx.Lock()
+	defer h.subscriberMtx.Unlock()
+
+	for _, sub := range h.subscribers {
+		sub.NewItemCreated.ChanIn() <- event
+	}
+}
+
+// wait blocks for a given amount of time.
+//
+// TODO(ffranr): add HashMailCourier quit signal?
+func (h *HashMailCourier) wait(ctx context.Context,
+	backoff time.Duration) error {
+
+	select {
+	case <-time.After(backoff):
+		return nil
+	case <-ctx.Done():
```

Comment:
Who would cancel this context? We may have to modify the `transferReceiverProof` logic to ensure that this context gets canceled if the system is shutting down :+1:

Reply:
Just to recap what we covered on the call (for my own clarity as much as anything else):

The the `wait` function has two call sites within `HashMailCourier`. And in both cases the passed in context comes from [this](https://github.com/ffranr/taro/blob/4bf1c0a13c7da1ab31f9b14fed87e08c9cf68309/tarofreighter/chain_porter.go#L466-L466) point in the `ChainPorter`.  The context is generated by the `ChainPorter`'s `ContextGuard` structure which cancels the context if the guard's associated `Quit` channel is closed or sent to.

The `ChainPorter`'s `Quit` channel is closed if `ChainPorter.Stop()` is called. And `ChainPorter.Stop()` is called by the exit procedure of the main server.

Reply:
It is, the context is created with `p.WithCtxQuitNoTimeout()`.